ISPN-6214 Initial cluster size

tristantarrant opened the pull request:
https://issues.jboss.org/browse/ISPN-6214

Also provides a fix for 

https://issues.jboss.org/browse/ISPN-6099

danberindei commented on core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:
```diff
@@ -462,6 +464,22 @@ public void waitForChannelToConnect() {
+   private void waitForInitialNodes() {
+      int initialClusterSize = configuration.transport().initialClusterSize();
+      if (initialClusterSize > 1) {
+         while (channel.getView().getMembers().size() < initialClusterSize) {
```
Should we have a timeout here?

tristantarrant replied:
Yes, we should probably use the same as the GMS timeout, wdyt ?

danberindei commented on core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:
```diff
@@ -227,7 +229,7 @@ protected void startJGroupsChannelIfNeeded() {
 
    @Override
-   public int getViewId() {
+   public int  getViewId() {
```
Unnecessary space

danberindei commented on core/src/test/java/org/infinispan/distribution/rehash/ConcurrentJoinTest.java:
```diff
@@ -26,7 +26,7 @@ void performRehashEvent(boolean offline) throws Exception {
 
       for (int i = 0; i < NUM_JOINERS; i++) {
-         EmbeddedCacheManager joinerManager = addClusterEnabledCacheManager(new TransportFlags().withFD(false));
+         EmbeddedCacheManager joinerManager = addClusterEnabledCacheManager(new TransportFlags().withFD(false).withPortRange(i));
```
You should do the same in `StateTransferFunctionalTest.testConcurrentStateTransfer`.

Although I'd rather see this solved in JGroups :)

wburns commented on core/src/main/java/org/infinispan/manager/DefaultCacheManager.java:
```diff
@@ -642,6 +642,7 @@ public void start() {
       if (globalConfiguration.security().authorization().enabled()) {
          globalConfiguration.security().authorization().principalRoleMapper().setContext(new PrincipalRoleMapperContextImpl(this));
       }
+      globalComponentRegistry.start();
```
We shouldn't need https://github.com/tristantarrant/infinispan/blob/ISPN-6214/initial_cluster_size/core/src/main/java/org/infinispan/manager/DefaultCacheManager.java#L610 any longer with this added.

wburns replied:
Also this change breaks a few tests.

# CacheManagerTest.testExceptionOnCacheManagerStop - simple fix
# CacheManagerNotifierTest.testMockViewChange - personally think this one might as well remove

wburns replied:
Still can remove the extra start method invocation for each cache being wired.

tristantarrant replied:
where ?

wburns replied:
https://github.com/tristantarrant/infinispan/blob/ISPN-6214/initial_cluster_size/core/src/main/java/org/infinispan/manager/DefaultCacheManager.java#L610

tristantarrant commented:
I think I have addressed all issues:
* dropped tests which don't mean anything anymore (CacheManagerNotifierTest.testMockViewChange)
* added a timeout check for the initial cluster based on the GMS timeout with test
* added the portRange "trick" to other tests so that concurrent channels don't use the same range of ports thus triggering ISPN-6099

wburns commented on core/src/test/java/org/infinispan/statetransfer/StateTransferFunctionalTest.java:
```diff
@@ -49,6 +49,7 @@
    private volatile int testCount = 0;
 
    private static final Log log = LogFactory.getLog(StateTransferFunctionalTest.class);
+   private int cacheManagerCount;
```
Unneeded

wburns commented on core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:
```diff
@@ -462,6 +475,26 @@ public void waitForChannelToConnect() {
+      int initialClusterSize = configuration.transport().initialClusterSize();
+      if (initialClusterSize > 1) {
+         GMS gms = (GMS) channel.getProtocolStack().findProtocol(GMS.class);
+         long timeout = gms != null ? gms.getJoinTimeout() : 15000;
```
I wonder if GMS timeout will be enough though.  GMS just starts a cluster if the timeout occurs, where as we will get an exception.  With a large cluster of nodes this will definitely not be enough time unless GMS was increased in the JGroups config.  Should this be a separate timeout maybe that starts out a bit higher?

tristantarrant replied:
I didn't really want to add yet another configuration property...

wburns replied:
I agree, just trying to think if this will be sufficient.

wburns commented:
Looks fine to me overall, just a couple minor things.

tristantarrant commented:
Removed the unnecessary gcr start(), @wburns

tristantarrant commented:
Complete with timeout, @wburns

tristantarrant commented:
I forgot server support. Adding it now.

tristantarrant commented:
Now complete with server support

danberindei commented on core/src/main/java/org/infinispan/configuration/parsing/Parser80.java:
```diff
@@ -803,6 +852,48 @@ private void parseTransport(XMLExtendedStreamReader reader, ConfigurationBuilder
                      createThreadPoolConfiguration(value, REMOTE_COMMAND_EXECUTOR));
                break;
             }
+            case EXTENDS:
```
Compilation error here. Also unnecessary...

danberindei commented on core/src/test/java/org/infinispan/remoting/transport/InitialClusterSizeTest.java:
```diff
@@ -0,0 +1,74 @@
+@CleanupAfterMethod
+public class InitialClusterSizeTest extends MultipleCacheManagersTest {
+   public static final int CLUSTER_SIZE = 4;
+   public static final int CLUSTER_TIMEOUT = 5;
```
For timeouts, I find it useful to put the time unit in the constant name, e.g. `CLUSTER_TIMEOUT_SECONDS`.

tristantarrant commented:
Fixde the compilation error since https://github.com/infinispan/infinispan/pull/4014 got merged

galderz commented:
Tested and integrated :)
